### PR TITLE
Subscriptions Shipping Packages Integration Points

### LINKED ISSUE #3241: Integration Tracking (WC Subscriptions): Show Shipping selector when checking out multiple subscriptions and persist those values
The second part of https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/3240

When you check out multiple subscription products with different billing schedules, you will create a subscription for each one of them, each with its own shipping method.
On the checkout, you can select which shipping method you will assign to each subscription, we should add this.

image

This has some design implications because we show shipping methods in a different part of the checkout block than what we have in the classic checkout.

### PR DESCRIPTION
**Pull request by mikejolley**
This PR creates a new `__experimentalOrderShippingPackages ` slot fill for shipping packages, and adds several integration points for subscriptions to add it's data ([the integration on the subscriptions side, including new endpoint data, can be seen in this PR](https://github.com/woocommerce/woocommerce-subscriptions/pull/3959)).

The initial package (renamed via a server side hook) is added via the SlotFill. The collapsible state of shipping methods is based on whether it's enabled (it is on cart, and isn't on checkout) and if there are more than 1 shipping packages, or more than 1 SlotFills present.

The client side subscriptions integration is also in this PR (within the cart checkout components code) and will need to be moved once components it requires are exported. Right now it's using `useStoreCart` and `Packages` from the `ShippingRateControl` package.

For the API endpoint integration, we needed to expose the schemas. Subscriptions is using this to create it's own array of shipping packages using the core schema.

This PR was created as an alternative to https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3672.

Closes #3241

**To test**
- You must be running subscriptions, and have this branch active and built https://github.com/woocommerce/woocommerce-subscriptions/pull/3959
- Create at least 2 subscription products with differing schedules, e.g. a monthly and a weekly subscription
- Add both to your cart
- See the shipping rates in the cart totals sidebar. They should be collapsible. 
- Go to checkout. You should see subscription shipping rates below the regular rates. They will not be collapsible

### CODE REVIEW

File: assets/js/blocks/cart-checkout/subscriptions-integration.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+				newPackages.push( recurringCartPackage );
+			} );
+		} );
+
+		return newPackages;
+	}, [ subscriptions ] );
+
+	const { selectShippingRate, selectedShippingRates } = useSelectShippingRate(
+		packages
+	);
```
Comment (senadir) [inline assets/js/blocks/cart-checkout/subscriptions-integration.js]: I'm not sure I understand `useSelectShippingRate`, does it allow selecting a rate from a single package, or a rate from different packages?

Also, is there any value of having it this high in the tree? can it be moved inside  `ShippingRatesControlPackage` or be supplied from `ExperimentalOrderShippingPackages` as a `fillProps`?

File: assets/js/blocks/cart-checkout/subscriptions-integration.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	ShippingRatesControlPackage,
+} from '@woocommerce/blocks-checkout';
+import { registerPlugin } from '@wordpress/plugins';
+import { useStoreCart, useSelectShippingRate } from '@woocommerce/base-hooks';
+import { useMemo } from '@wordpress/element';
+
+const RenderSubscriptionPackages = () => {
+	const {
+		extensions: { subscriptions },
+	} = useStoreCart();
```
Comment (senadir) [inline assets/js/blocks/cart-checkout/subscriptions-integration.js]: I didn't fully test this pattern yet, but I think we can have `RenderSubscriptionPackages` in a higher level SlotFill (CartCheckout maybe?) that passes down `cart` data instead of using useStoreCart.
If it's not possible, we can expose a new hook that only pass down relevant data we need, but we would need to remove `fillProps` to not have to implementation of the same thing.

File: assets/js/blocks/cart-checkout/subscriptions-integration.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		Object.values( subscriptions ).forEach( ( recurringCart ) => {
+			const recurringCartPackages = recurringCart.shipping_rates || [];
+
+			recurringCartPackages.forEach( ( recurringCartPackage ) => {
+				newPackages.push( recurringCartPackage );
+			} );
+		} );
+
+		return newPackages;
+	}, [ subscriptions ] );
```
Comment (senadir) [inline assets/js/blocks/cart-checkout/subscriptions-integration.js]: This felt really hard for me to read, it can be simplified to this
```suggestion
const packages = useMemo(
		() =>
			Object.values( subscriptions )
				.map( ( recurringCart ) => recurringCart.shipping_rates )
				.filter( Boolean )
				.flat(),
		[ subscriptions ]
	);
```

File: assets/js/blocks/cart-checkout/subscriptions-integration.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+	return packages.map( ( { package_id: packageId, ...packageData } ) => (
+		<ExperimentalOrderShippingPackages key={ packageId }>
+			<SubscriptionPackage
+				key={ packageId }
+				packageData={ packageData }
+				onSelectRate={ ( newShippingRate ) => {
+					selectShippingRate( newShippingRate, packageId );
+				} }
+				selected={ selectedShippingRates[ packageId ] }
```
Comment (senadir) [inline assets/js/blocks/cart-checkout/subscriptions-integration.js]: Nothing hererelies on external data and all can be performed internally, given that we're publishing a very opinioned shipping package selector, would it make sense to abstract this part?

File: packages/checkout/order-shipping-packages/index.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	className,
+	collapsibleWhenMultiple,
+	collapsible,
+	noResultsMessage,
+	renderOption,
+} ) {
+	const { fills } = useSlot( slotName );
+	const hasMultiplePackages = fills.length > 1;
+	return (
+		<SlotComponent
```
Comment (senadir) [inline packages/checkout/order-shipping-packages/index.js]: A good pattern (from GB) is to name this `${slotName}Slot` so here it would be `OrderShippingPackagesSlot`.

File: packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js [hunk 6]
```diff
@@ -87,22 +104,54 @@ const ShippingRatesControl = ( {
...
+				collapsibleWhenMultiple={ collapsibleWhenMultiple }
 				noResultsMessage={ noResultsMessage }
 				renderOption={ renderOption }
-				shippingRates={ shippingRates }
+			/>
+			<Packages
+				packages={ shippingRates }
+				noResultsMessage={ noResultsMessage }
 			/>
 		</LoadingMask>
```
Comment (senadir) [inline packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js:117]: I'm not sure this code should live in our public package, or if it does make sense to be here. Seeing as this is calling a slotFill, I'd prefer if it's in our internal code.

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 1 above)
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: It manages selected state across multiple packages, as it did prior to this PR. In this context we are using it to manage state of the multiple packages subscriptions can return, separate from the multiple packages core can return for regular items.

File: packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js [hunk 7]
```diff
@@ -87,22 +104,54 @@ const ShippingRatesControl = ( {
...
+			<Package
+				key={ packageId }
+				packageData={ packageData }
+				onSelectRate={ ( newShippingRate ) => {
+					selectShippingRate( newShippingRate, packageId );
+				} }
+				selected={ selectedShippingRates[ packageId ] }
+			/>
+		</ExperimentalOrderShippingPackages>
+	) );
```
Comment (senadir) [inline packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js]: I'm not sure this code should live in our public package, or if it does make sense to be here. Seeing as this is calling a slotFill, I'd prefer if it's in our internal code.

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 1 above)
Comment (senadir) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: Would it make sense to limit it to a single package?

Comment (mikejolley) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: The selected rates I think are exposed in other places... If we do it per package that will make that more difficult?

Side note, I actually wanted to remove selected methods state from this hook competely but had some pushback earlier in the week. Happy to refactor further - just say so.

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 2 above)
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: Are there use cases for exposing full cart hook? What is dangerous to expose exactly?

File: packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js (hunk 7 above)
Comment (mikejolley) [reply packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js]: It wasn't prior to some of the other refactors. Does the entire control need to be in this public package? If we move it, where else is suitable?

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js [hunk 8]
```diff
@@ -2,77 +2,91 @@
...
-					return obj;
-				}, {} ),
-		[ shippingRates ]
-	);
 
-	const [ selectedShippingRates, setSelectedShipping ] = useState(
-		derivedSelectedRates
+	// Selected rates are stored in state.
+	const [ selectedShippingRates, setSelectedShipping ] = useState( () =>
+		deriveSelectedRates( shippingRates )
```
Comment (nerrad) [inline assets/js/base/hooks/shipping/use-select-shipping-rate.js]: nice little refactor here to remove the need for `useMemo` 👏

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js [hunk 9]
```diff
@@ -2,77 +2,91 @@
...
 
 /**
  * Internal dependencies
  */
 import { useThrowError } from '../use-throw-error';
 
+/**
+ * Selected rates are derived by looping over the shipping rates. The selected rates will have a true selected prop.
+ *
+ * @param {Array} shippingRates of shipping rates.
```
Comment (nerrad) [inline assets/js/base/hooks/shipping/use-select-shipping-rate.js]: Looks like to me this is returning an Object with the following shape?
```js
const selectedRates = { [packageId:string]: selected:boolean }
```

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js [hunk 10]
```diff
@@ -2,77 +2,91 @@
...
+				[ packageId ]: newShippingRate,
+			} );
+			selectShippingRate( newShippingRate, packageId ).catch(
+				( error ) => {
+					// we throw this error because an error on selecting a rate is problematic.
+					throwError( error );
+				}
+			);
+		},
+		[ selectedShippingRates, throwError, selectShippingRate ]
```
Comment (nerrad) [inline assets/js/base/hooks/shipping/use-select-shipping-rate.js]: `throwError` is going to be a new function on every render calc I think, so the `useCallback` here is unnecessary memoization it looks like. We could modify the function returned from `throwError` so it's memoized in a `useCallback`?

File: assets/js/blocks/cart-checkout/checkout/form/shipping-options-step.js [hunk 11]
```diff
@@ -73,7 +74,7 @@ const ShippingOptionsStep = () => {
 			className="wc-block-checkout__shipping-option"
 			title={ __( 'Shipping options', 'woo-gutenberg-products-block' ) }
 			description={
-				getShippingRatesRateCount( shippingRates ) > 1
+				rateCount > 1
```
Comment (nerrad) [inline assets/js/blocks/cart-checkout/checkout/form/shipping-options-step.js]: Nit: I'm curious why this change? If anything we should remove the assigning of a `packageCount` and invoke `getShippingRatesPackageCount` directly. I generally think it's good practice to avoid assigning variables unless the value is used in more than one place.

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js (hunk 8 above)
Comment (nerrad) [reply assets/js/base/hooks/shipping/use-select-shipping-rate.js]: Actually, in general this is a nice refactor/cleanup of the `useSelectShippingRate` hook!

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 2 above)
Comment (nerrad) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: As soon as we expose lower level hooks etc, we increase the surface area of maintenance and more implementation details. It's preferable to have more control over the data we expose to extensions via props or a special public interface hook(s).

For instance `useStoreCart` exposes `receiveCart` which would allow extensions to override the cart in the state. I'm not sure we want to do that.

My comments here apply to exposing `useSelectShippingRate` hook as well.

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 4 above)
Comment (nerrad) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: Ya I concur with Nadir here. It looks like the only subscriptions specific thing here is getting subscriptions from the extensions property and converting that to a packages array. Could we just have extensions register their own callback for modifying the packages when additional shipping areas are created?  So in this case in whatever interface is provided Subscriptions would just provide this as a callback for packages:

```js
const getPackages = ( cart ) => {
	const { subscriptions } = cart.extensions;
	return Object.values( subscriptions )
		.map( ( recurringCart ) => recurringCart.shipping_rates )
		.filter( Boolean )
		.flat();
};
```

Notice the callback would receive the cart object so the extension could use that to derive its packages from (like what you are doing currently in this code).

File: packages/checkout/shipping/shipping-rates-control/package-rates.js [hunk 12]
```diff
@@ -6,25 +6,32 @@ import RadioControl, {
...
+import { renderPackageRateOption } from './render-package-rate-option';
+
+const PackageRates = ( {
 	className,
 	noResultsMessage,
-	onChange,
-	options,
-	renderOption,
+	onSelectRate,
+	rates,
```
Comment (nerrad) [inline packages/checkout/shipping/shipping-rates-control/package-rates.js:18]: nice prop renaming here for clarity 👍

File: src/Domain/Services/ExtendRestApi.php [hunk 13]
```diff
@@ -29,14 +38,26 @@ class ExtendRestApi {
 	/**
 	 * Constructor
 	 *
+	 * @param Container  $container An instance of the container class.
 	 * @param Package    $package An instance of the package class.
 	 * @param Formatters $formatters An instance of the formatters class.
 	 */
-	public function __construct( Package $package, Formatters $formatters ) {
+	public function __construct( Container $container, Package $package, Formatters $formatters ) {
```
Comment (nerrad) [inline src/Domain/Services/ExtendRestApi.php]: I don't think we should pass in the container. It works, but the proper pattern (I think) would be to pass in a schema repository class that is used to retrieve the schema. It reduces the scope of making the entire container available to this class. 

But I'm not sure how the schema repository class would then be created/implemented. It's kind of a chicken/egg scenario at risk of creating circular dependencies because ideally `SchemaController` would be passed into `ExtendRestApi` but we have `ExtendRestAPI` passed into the `SchemaController` AND each of the schemas 🤔 .

Until we can come up with a better solution, an interim approach for now (even though it's also an antipattern - service locator) would be to just call `Package::container()` directly in the new `getSchema` method here. And we can add a todo to improve this. That way we're not creating a signature for the constructor that we might change significantly down the road.

**The larger question here**, is what need is there for a `get_schema` method on this class (I know it's currently implemented in the companion Subscriptions extension pr)? Shouldn't any code needing schemas use the `SchemaController` directly? I'm pretty inclined to revisit this approach as I'm pretty wary about the cyclical dependency this creates.

Comment (nerrad) [reply src/Domain/Services/ExtendRestApi.php]: Speaking of which, I wonder if we should make this class final?

Comment (mikejolley) [reply src/Domain/Services/ExtendRestApi.php]: I did it this way because of the reason you mentioned; circular dependency. Subs needs access to our schema definitions via extend, but the schema definitions themselves are passed the extend class. I wasn't sure if there was another way around this.

Comment (nerrad) [reply src/Domain/Services/ExtendRestApi.php]: Ya I get that. Still to avoid creating a signature that's hard to walk back from, I think for now it'd be wise to do a couple things here:

- Make the `ExtendRestAPI` class final.
- Directly get the container from `Package:container` for now in `get_schema` and add a todo as tech debt to revisit in the future. 

Thoughts?

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js (hunk 10 above)
Comment (mikejolley) [reply assets/js/base/hooks/shipping/use-select-shipping-rate.js]: Is this what you meant? https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3687/commits/41ca1871d7bfe78c3e7617f6b70fec37fcd24a5b setState is stable between renders, but it was being wrapped in an anonymous function.

File: assets/js/blocks/cart-checkout/checkout/form/shipping-options-step.js (hunk 11 above)
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/checkout/form/shipping-options-step.js]: I changed this as a personal preference; I find it easier to follow when we have constants in our ternaries than functions, especially when there are things with similar names (in this case we have a `getShippingRatesPackageCount` and a `getShippingRatesRateCount` which is easy to misread inline).

If it's problematic I can change it back.

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js (hunk 9 above)
Comment (mikejolley) [reply assets/js/base/hooks/shipping/use-select-shipping-rate.js]: I'm not sure if it's possible to represent the dynamic key in typedef? In added an line note about the structure of the returned object however. https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3687/commits/8eb6d5bf5e3bc7e9f4e1731aa4822ec4892849b7

File: assets/js/base/hooks/shipping/use-select-shipping-rate.js (hunk 8 above)
Comment (mikejolley) [reply assets/js/base/hooks/shipping/use-select-shipping-rate.js]: You can thank epic react :) Thanks!

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 4 above)
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: Is there any prior art for this suggestion?

File: src/Domain/Services/ExtendRestApi.php (hunk 13 above)
Comment (mikejolley) [reply src/Domain/Services/ExtendRestApi.php]: I think we should maybe think about how, or if, Extend should provide Schema to extensions, because if we include `get_schema ` for extensions we cannot back out of that contract.

@senadir did you have any thoughts on this?

I wonder if we could rejig how extend works so rather than it being passed to schema classes, it's invoked when schema classes are used by the routes.

**Comment by opr** [comment]
I've tested this and I can confirm that rates for each package are shown both in the cart and checkout blocks, they look good and I can switch rates and see the price update when the _initial package_ rate is changed. 🥳

One thing I spotted in the cart block is that the border for the focused radio button is cut off:
Screenshot 2021-01-22 at 17 26 17

I can't comment too much on the code changes, because I'm still unfamiliar with the areas this PR touches, but I am a fan of some of the refactoring you've done and renaming variables so their purpose is clearer 👌🏻

I'm 👀 the code discussions in this PR though and will input if I have anything to add.

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 1 above)
Comment (senadir) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: > The selected rates I think are exposed in other places... If we do it per package that will make that more difficult?

I assume selected rates are needed for express payments? can't they be derived from cart response?

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 4 above)
Comment (senadir) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: > Is there any prior art for this suggestion?

For my comment or for Darren's comment?

>  Could we just have extensions register their own callback for modifying the packages when additional shipping areas are created?

This part is not obvious to me, do you mean having a callback on `ShippingRatesControlPackage` like this:

```js
const SubscriptionPackage = ( props ) => {
	const getPackages = useCallback( ( cart ) => {
		const { subscriptions } = cart.extensions;
		return Object.values( subscriptions )
			.map( ( recurringCart ) => recurringCart.shipping_rates )
			.filter( Boolean )
			.flat();
	}, [] );
	return (
		<ShippingRatesControlPackage { ...props } packageData={ getPackages } />
	);
};
```

File: packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js (hunk 7 above)
Comment (senadir) [reply packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js]: I think we moved it while trying to fix some circular dependency issue but we can return it if we're not exposing it to extensions if not, the least thing we can do is move back the Slot part back to `base/components/cart-checkout`.

File: src/Domain/Services/ExtendRestApi.php (hunk 13 above)
Comment (senadir) [reply src/Domain/Services/ExtendRestApi.php]: > I think we should maybe think about how, or if, Extend should provide Schema to extensions, because if we include get_schema for extensions we cannot back out of that contract.

I believe we discussed this before; what value do we gain by exposing our schema to extensions? We're neither validating the developer schema nor responsible for rendering it.

One argument for this is that having the same schema would make code reuse much easier but would put a higher maintenance cost for us, giving that we're committing to our REST API shape (and therefore schema), this should be fine.

> I wonder if we could rejig how extend works so rather than it being passed to schema classes, it's invoked when schema classes are used by the routes.

I don't get this part? This is how it works right now, extend hook into `get_properties` and `get_item_response` which are evaluated when the route actually use them.

Comment (mikejolley) [reply src/Domain/Services/ExtendRestApi.php]: @senadir Right now extend works within each schema `get_item_response ` - I was suggesting considering moving the extend logic to the Route where `get_item_response` is called.

Comment (senadir) [reply src/Domain/Services/ExtendRestApi.php]: So something like this?
```php
return rest_ensure_response( array_merge( $this->schema->get_item_response( $cart ), $this->extend->get_extended_schema( self::IDENTIFIER ) ) );
```

Comment (mikejolley) [reply src/Domain/Services/ExtendRestApi.php]: @senadir I'd add a utility to merge, but yes, that. Routes are not something we'll pass to extensions, but schema we may, so this avoids circular dependencies  _because schema will have no dependencies._

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 1 above)
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: @senadir This is what it took to move it to package level and remove from the subscriptions integration. We track state at package level, but the select shipping rates hook now has no state so that it can simply make requests and determine if a request is being made. Select rates is derived where needed.

https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3687/commits/5b7f860113fb7ad3d6741edfabf00e562d57253a

File: assets/js/blocks/cart-checkout/subscriptions-integration.js (hunk 4 above)
Comment (mikejolley) [reply assets/js/blocks/cart-checkout/subscriptions-integration.js]: I think we can combine this feedback with https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3687#discussion_r562164364

File: src/Domain/Services/ExtendRestApi.php (hunk 13 above)
Comment (mikejolley) [reply src/Domain/Services/ExtendRestApi.php]: @senadir Not as straightforward as hoped as some schemas use their own schema to derive props. Lets leave as is—I'll look for alternative solution.

Comment (mikejolley) [reply src/Domain/Services/ExtendRestApi.php]: I've reverted the change and used it directly in subs https://github.com/woocommerce/woocommerce-subscriptions/pull/3959/commits/ff158451e38e997dfd7bbc10020428940ba6a080

File: assets/js/base/context/cart-checkout/shipping/index.js [hunk 14]
```diff
@@ -116,6 +83,19 @@ export const ShippingDataProvider = ( { children } ) => {
...
+	);
+	useEffect( () => {
+		const derivedSelectedRates = deriveSelectedShippingRates(
+			shippingRates
+		);
+		if ( ! isShallowEqual( selectedRates.current, derivedSelectedRates ) ) {
+			selectedRates.current = derivedSelectedRates;
+		}
+	}, [ shippingRates ] );
+
```
Comment (senadir) [inline assets/js/base/context/cart-checkout/shipping/index.js:98]: I think we do here with `usePrevious` hook

File: assets/js/blocks/cart-checkout/subscriptions-integration.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+				packageId={ packageId }
+				packageData={ packageData }
+			/>
+		</ExperimentalOrderShippingPackages>
+	) );
+};
+
+const SubscriptionPackage = ( props ) => {
+	return <ShippingRatesControlPackage { ...props } />;
+};
```
Comment (senadir) [inline assets/js/blocks/cart-checkout/subscriptions-integration.js]: This seems to be useless now and I will delete it once I move the code if you don't mind Mike.

File: packages/checkout/order-shipping-packages/index.js [hunk 16]
```diff
@@ -0,0 +1,67 @@
+/**
+ * External dependencies
+ */
+import {
+	createSlotFill,
+	__experimentalUseSlot as useSlot,
```
Comment (senadir) [inline packages/checkout/order-shipping-packages/index.js:10]: Since we can't get away without this hook, in a follow-up, I'm going to do a defensive check in case it's no longer `experimental` or if it's removed.

File: packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js [hunk 17]
```diff
@@ -14,9 +14,25 @@ import {
...
+ *
+ * @param {Object} props Incoming props.
+ * @param {Array} props.shippingRates Array of packages containing shipping rates.
+ * @param {boolean} props.shippingRatesLoading True when rates are being loaded.
+ * @param {string} props.className Class name for package rates.
+ * @param {boolean} props.collapsibleWhenMultiple If true, when multiple packages are rendered they can be toggled open and closed.
+ * @param {React.ReactElement} props.noResultsMessage Rendered when there are no packages.
+ * @param {Function} props.renderOption Function to render a shipping rate.
+ */
 const ShippingRatesControl = ( {
```
Comment (senadir) [inline packages/checkout/shipping/shipping-rates-control/shipping-rates-control.js:39]: In a follow up, this component, with `TotalsShipping` and `ShippingRateSelector` need to be moved back to our base components folder, they're too coupled with our base implementation to be exported, if anything, we should only export `Package`.

**Review by senadir — APPROVED** [review]
Sink it.
